flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1619,6 +1616,34 @@ public void testTaskAvoidHangingAfterSnapshotStateThrownException() throws Excep
         }
     }
 
+    @Test
+    public void testCleanUpResourcesWhenFailingDuringInit() throws Exception {
+        // given: Configured SourceStreamTask with source which fails during initialization.
+        Configuration taskManagerConfig = new Configuration();
+        taskManagerConfig.setString(STATE_BACKEND, TestMemoryStateBackendFactory.class.getName());
+
+        StreamConfig cfg = new StreamConfig(new Configuration());
+        cfg.setStateKeySerializer(mock(TypeSerializer.class));
+        cfg.setOperatorID(new OperatorID(4712L, 43L));
+
+        cfg.setStreamOperator(new TestStreamSource<>(new InitFailedSource()));
+        cfg.setTimeCharacteristic(TimeCharacteristic.ProcessingTime);
+
+        try (NettyShuffleEnvironment shuffleEnv = new NettyShuffleEnvironmentBuilder().build()) {
+            Task task = createTask(SourceStreamTask.class, shuffleEnv, cfg, taskManagerConfig);
+
+            // when: Task starts.
+            task.startTaskThread();
+
+            // wait for clean termination.
+            task.getExecutingThread().join();
```

Comment:
I would suggest to simplify and implement this test on a lower level in more unit test style, bypassing `Task` class, and without spawning extra threads. You could check how it can be done for example in `StreamTaskTest#testNotifyCheckpointOnClosedOperator`. Something like:

```
StreamTaskMailboxTestHarnessBuilder<Integer> builder =
        new StreamTaskMailboxTestHarnessBuilder<>(
                        OneInputStreamTask::new, BasicTypeInfo.INT_TYPE_INFO)
                .addInput(BasicTypeInfo.INT_TYPE_INFO);
StreamTaskMailboxTestHarness<Integer> harness =
        builder.setupOutputForSingletonOperatorChain(new YourTestOperator()).build();
harness.streamTask.restore();
```